modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -2328,19 +2345,25 @@ private CompletableFuture<Boolean> ensureReplicaIsPrimary(ReplicaRequest request
         }
 
         if (expectedTerm != null) {
-            return raftClient.refreshAndGetLeaderWithTerm()
-                    .thenCompose(replicaAndTerm -> {
-                                long currentTerm = replicaAndTerm.term();
-
-                                if (expectedTerm == currentTerm) {
-                                    return completedFuture(null);
+            return placementDriver.getPrimaryReplica(replicationGroupId, hybridClock.now().addPhysicalTime(HybridTimestamp.CLOCK_SKEW))
+                    .thenCompose(primaryReplica -> {
+                                long currentEnlistmentConsistencyToken = primaryReplica.getStartTime().longValue();
+
+                                if (expectedTerm.equals(currentEnlistmentConsistencyToken)) {
+                                    if (primaryReplica.getExpirationTime().before(hybridClock.now())) {
+                                        return failedFuture(
+                                                new PrimaryReplicaMissException(expectedTerm, currentEnlistmentConsistencyToken));
+                                    } else {
+                                        return completedFuture(null);
+                                    }
                                 } else {
-                                    return failedFuture(new PrimaryReplicaMissException(expectedTerm, currentTerm));
+                                    return failedFuture(new PrimaryReplicaMissException(expectedTerm, currentEnlistmentConsistencyToken));
                                 }
                             }
                     );
         } else if (request instanceof ReadOnlyReplicaRequest || request instanceof ReplicaSafeTimeSyncRequest) {
-            return raftClient.refreshAndGetLeaderWithTerm().thenApply(replicaAndTerm -> isLocalPeer(replicaAndTerm.leader()));
+            return placementDriver.getPrimaryReplica(replicationGroupId, hybridClock.now().addPhysicalTime(HybridTimestamp.CLOCK_SKEW))
```

Comment:
Seems that there are no usages of `getPrimaryReplica` without adding `CLOCK_SKEW` to the argument, and this clock skew be always necessary. Shouldn't we add this clock skew to time comparison inside of this method?

Reply:
CLOCK_SKEW addition moved to getPrimaryReplica()